### PR TITLE
Fix CG+<service/route> associated plugins

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:

Make KongPlugins applied to both a KongConsumerGroup and Ingress or Service produce a plugin bound to both a consumer group and route (or service).

**Which issue this PR fixes**:

The test portion added by this PR currently fails. Instead of creating the plugin described above, this configuration creates a plugin attached to the route alone, so the final added case (accessing the route with a consumer _not_ in the associated consumer group) does see the header that should only appear for consumers in the group.

**Special notes for your reviewer**:

I don't have great explanation of the logic in `GetCombinations()`, but bozo mode fix approach is to make its CG handling similar to its consumer handling, since consumers do behave as expected for plugins mapped to both. I'm not sure why consumers need this while route+service plugins apparently don't, and only need the simple logic to add a relation if requested.

History suggests the split is quite old (b65453a8d84cf8645ef1e980b50a4663d7ca2db9) and I don't want to dig into the maths version of the code to explain why this matters for a consumer+route and CG+route plugin but not a route+service plugin. Maybe you do? IDK, it works, bozo maths succeeds.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/util/relations.go [hunk 1]
```diff
@@ -35,10 +37,31 @@ func (relations *ForeignRelations) GetCombinations() []Rel {
 				cartesianProduct = append(cartesianProduct, Rel{Consumer: consumer})
 			}
 		}
-	} else {
-		for _, consumerGroup := range relations.ConsumerGroup {
-			cartesianProduct = append(cartesianProduct, Rel{ConsumerGroup: consumerGroup})
+	} else if len(relations.ConsumerGroup) > 0 {
+		groups := relations.ConsumerGroup
+		if len(relations.Route)+len(relations.Service) > 0 {
```
Comment (randmonkey) [inline internal/util/relations.go:42]: Could we use `len(relations.Route) > 0 || len(relations.Service) > 0` to let this be more intuitive?

File: internal/util/relations.go [hunk 2]
```diff
@@ -35,10 +37,31 @@ func (relations *ForeignRelations) GetCombinations() []Rel {
...
+			for _, service := range relations.Service {
+				for _, group := range groups {
+					cartesianProduct = append(cartesianProduct, Rel{
```
Comment (czeslavo) [inline internal/util/relations.go:45]: Please make sure the new cases are covered in the existing `TestGetCombinations` unit test.

**Comment by pmalek** [comment]
I've suggested a small refactor which will greatly speed up the generation of relations: https://github.com/Kong/kubernetes-ingress-controller/pull/6152

That PR has this set as target so feel free to merge it. It also adds a CG test case.

File: CHANGELOG.md [hunk 3]
```diff
@@ -224,6 +224,12 @@ Adding a new version? You'll need three changes:
   [#6079](https://github.com/Kong/kubernetes-ingress-controller/pull/6079)
 - Fixed KIC non leaders correctly getting up to date Admin API addresses by not
   requiring leader election for the related controller.
+  [6126](https://github.com/Kong/kubernetes-ingress-controller/pull/6126)
+- Plugins attached to both a KongConsumerGroup and a route-like resource or
+  Service now properly generate a plugin attached to both a Kong consumer group
+  and route or service. Previously, these incorrectly generated plugins
+  attached to the route or service only.
+  [6132](https://github.com/Kong/kubernetes-ingress-controller/pull/6132)
```
Comment (randmonkey) [inline CHANGELOG.md:232]: duplicated changelog entries here
```suggestion

```

File: CHANGELOG.md [hunk 4]
```diff
@@ -224,6 +224,12 @@ Adding a new version? You'll need three changes:
   [#6079](https://github.com/Kong/kubernetes-ingress-controller/pull/6079)
 - Fixed KIC non leaders correctly getting up to date Admin API addresses by not
   requiring leader election for the related controller.
+  [6126](https://github.com/Kong/kubernetes-ingress-controller/pull/6126)
```
Comment (randmonkey) [inline CHANGELOG.md:227]:
```suggestion
  [#6126](https://github.com/Kong/kubernetes-ingress-controller/pull/6126)
```

File: CHANGELOG.md (hunk 3 above)
Comment (randmonkey) [inline CHANGELOG.md:232]:
```suggestion
  [#6132](https://github.com/Kong/kubernetes-ingress-controller/pull/6132)
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -224,6 +224,12 @@ Adding a new version? You'll need three changes:
...
+- Plugins attached to both a KongConsumerGroup and a route-like resource or
+  Service now properly generate a plugin attached to both a Kong consumer group
+  and route or service. Previously, these incorrectly generated plugins
+  attached to the route or service only.
+  [6132](https://github.com/Kong/kubernetes-ingress-controller/pull/6132)
   [#6126](https://github.com/Kong/kubernetes-ingress-controller/pull/6126)
```
Comment (randmonkey) [inline CHANGELOG.md:233]:
```suggestion
```
Looks like some incorrect changes happened on the changelog entries

**Review by randmonkey — DISMISSED** [review]

**Review by tao12345666333 — APPROVED** [review]

**Review by tao12345666333 — APPROVED** [review]

**Comment by pmalek** [comment]
Manual  backport in https://github.com/Kong/kubernetes-ingress-controller/pull/6160

**Pull request merged**